docs/index_while_building.md

```diff
@@ -0,0 +1,102 @@
+building, `swift` and `clang` compilers use a global index cache internally to
+preserve performance. Finally, to integrate with remote caching, actions copy
+data into `bazel-out`.
+
+### rules_swift workers
+
+Workers are extended to use a global index internally. Then, it writes records
+and units into the remote cache by copying material from the worker's global
+index into `bazel-out`. 
+
+### clang compilation
+
+The native rules don't have an extra output for indexstore. These rules are
+updated to use the global index and in M3 remote caching for clang compilation
+is added.
+
+
+#### Incrementally importing remotely compiled indexes
+
+In order to import indexs into Xcode incrementally, a program is invoked to
+"import" remotely compiled indexes. This may be an aspect that runs during the
+Xcode build in order to pass an output file map to index import, or an ad-hoc
+program picking up output maps from build events.
+
+By default clang and swift seem to write absolute paths into the Index. If this
+is invariant continues to hold, it needs to be accounted for in Bazel. e.g.
+Unless all source code and compilation resided at the same directory remotely
+and locally there will likely need to be a `remap`ing of paths. The previously
+mentioned importing program will be smart enough to govern this - and import
+only the files that were recompiled.
+
+### Local optimization - sans index-import
+
+Initially the feature will work the same locally as it does remotely.
+
+However, one major shortcut exists locally to remove index-import. Xcode should
+read directly from Bazel's global index store. In other words, instead of
+writing to the per module index, and copying to `bazel-out` for remote caching,
+Xcode should pull right from that index. A "out of band" sentenial value can noop
+the code path for `bazel-out`.
+
+In order to remove the requirement to "index-import" bazel indexes, there are
+special considerations in the way that Xcode and Bazel invoke compilation and
+index-while-building abilities. Specifically, the file paths that Xcode and
+Bazel use to address index records must be aligned. This may be achievable by
+just aligning outputs between Xcode and Bazel: for example in the build system
+xcbuild ( e.g. XCBuildKit ) protocol messages, in xcconfig by means of the
+`.xcspec` features, or by plugin.
+
+
+## Task roadmap 
+
+
+## Determine direction forward with prototypes
+1. Consider possibilities of improving performance of "Index while building"
+There are a couple avenues here:
+
+- solution that works with remote caching
+
+- A solution that only runs locally. This is the status quo of objc right
+now: it doesn't work with remote caching
+
+2. Disable "Index while building" and correspondingly index-import
+Need to consider the user facing impact  MDX-3735
+
+
+## Land index while building V2 - first pass
+
+1. Propose changes to `rules_swift` for compilation
+2. Implement an aspect to quickly pull remotely built indexes
```

Comment:
In rules_ios, we currently look at BEP output to determine which indexstores were modified: https://github.com/bazel-ios/rules_ios/blob/master/tools/xcodeproj_shims/installers/indexstores.sh#L9

Introduced by this commit https://github.com/bazel-ios/rules_ios/commit/609d9e5a68cdcb64bf61a326906b096790d63489

Reply:
Initially we might able to use that that with the current rev of https://github.com/bazelbuild/rules_swift/pull/567 without more effort 👍 .

Longer term the aspect can import indexes while we're waiting for deps to compile - VS waiting until the end. If we go with a BEP reader that is running during the build like in xcbuildkit we might be able tp piggy back on that https://github.com/jerrymarino/xcbuildkit/blob/master/Examples/BazelBuildService/main.swift#L19